pksclient.8
.\"
.\" $Id: pksclient.8,v 1.1.1.1 2002/09/04 20:48:53 dtype Exp $
.\"
.\" Copyright (c) 1996, 1997, 1998, 1999, Marc Horowitz.  All rights reserved.
.\" See the LICENSE file in the release for redistribution information.
.\"
.TH PKSCLIENT 8 "17 November 1996"
.SH NAME
pksclient \- Public Key Server Client
.SH SYNOPSIS
.B pksclient
/db/path cmd [args ...]
.SH DESCRIPTION
pksclient is a command line interface which allows the administrator
to perform key server operations directly instead of through the
daemon.  The software will use locking and transaction semantics
unless you specify otherwise.  Never do this if another process has
the database open.  Once the command is completed,
.B pksclient
will attempt to checkpoint the database and remove any excess log
files.

Each command takes the path to the directory which contains the
database files, a command name, and possibly a list of arguments to
the command.  Some commands take an optional flags argument.  Flags
are single letters as described for each command, concatenated into a
single argument.

.TP 15
.B pksclient /db/path create [num_files]
Create an empty database.  If a database exists in the directory, it
is destroyed.  If
.B num_files
is specified, then the key database will be split into that many
files.  If it is not specified, then three key database files will be
created.
.TP
.B pksclient /db/path recover
Recover an inconsistent database.  This is equivalent to the
.B db_recover(8)
command, but it uses a larger cache for better performance.
.TP
.B pksclient /db/path add filename [flags]
Add a keyring to the database.  The
.B filename
may refer to a keyring file (.pgp format), or an ASCII-armored keyring
(.asc format).  If the
.B 'n'
flag is specified, then an incremental for the new keyring relative to
the database is generated to stdout.  If the
.B 'd'
flag is specified, then the disabled flag will not be stripped from
the input file.  This is useful if you are initializing the database
for the first time with a keyring from another key server which
includes disabled keys. If the
.B 't'
flag is specified, the operation will take place without logging and
transactions.  This is faster, but less safe.
.TP
.B pksclient /db/path get userid [flags]
An ASCII-armored keyring containing all the keys matching the 
.B userid
is printed to stdout.  If the
.B 'e'
flag is specified, then the key's user id must be an exact
case-insensitive substring of the 
.B userid
argument.  If the
.B 'a'
flag is specified, then the 
.B userid
argument is ignored, and all keys in the database are returned.  If
the
.B 'b'
flag is specified, then the output keyring will be in binary format
instead of ASCII-armor format.  If the
.B 'i'
flag is specified, then errors will be ignored.  This is used when
recovering from a corrupt database.  If the
.B 'd'
flag is specified, then disabled keys will be returned.  If the
.B 's'
flag is specified, the selected keys will be output unsorted to
stdout.  This flag implies the
.B 'b'
flag.  If the
.B 't'
flag is specified, the operation will take place without logging and
transactions.  This is faster, but less safe.
.TP
.B pksclient /db/path index userid [flags]
An index listing for all the keys matching the 
.B userid
is printed to stdout.  If the 
.B 'v'
flag is specified, then signatures are included in the output.  If the
.B 'f'
flag is specified, then the key fingerprint is included in the output.
If the 
.B 'e'
flag is specified, then the key's user id must be an exact
case-insensitive substring of the 
.B userid
argument.  If the 
.B 'a'
flag is specified, then the
.B userid
argument is ignored, and all keys in the database are indexed.  If the
.B 'i'
flag is specified, then errors will be ignored.  This is used when
recovering from a corrupt database.  If the 
.B 'd'
flag is specified, then disabled keys will be returned.  If the
.B 's'
flag is specified, the index will be output unsorted to stdout.  If the
.B 't'
flag is specified, the operation will take place without logging and
transactions.  This is faster, but less safe.
.TP
.B pksclient /db/path since time [flags]
An ASCII-armored keyring containing all the keys added to the database
or changed since the unix timestamp 
.B since
is printed to stdout.  The timestamp that the
database was last modified is printed to stderr.  If the
.B 'b'
flag is specified, then the output keyring will be in binary format
instead of ASCII-armor format. If the
.B 'r'
flag is specified, then the time given is taken as the number of seconds
the in the past the dump should start (
.I since 86400 r
is thus equivalent to the
.I LAST 1
mail command).  If the
.B 't'
flag is specified, the operation will take place without logging and
transactions.  This is faster, but less safe.
.TP
.B pksclient /db/path delete userid [flags]
All keys matching the
.B userid
are deleted from the database.  If the
.B 't'
flag is specified, the operation will take place without logging and
transactions.  This is faster, but less safe.
.TP
.B pksclient /db/path disable userid [flags]
All keys matching the
.B userid
have the disabled flag set.  If the
.B 'c'
flag is specified, then the flag is cleared instead of set.  If the
.B 't'
flag is specified, the operation will take place without logging and
transactions.  This is faster, but less safe.
.SH SEE ALSO
pks-intro(8), pksd(8)
.SH AUTHOR
Marc Horowitz, Massachusetts Institute of Technology


NEWS
For general information about the keyserver, visit
<http://sourceforge.net/projects/pks/>.

***** Changes in version 0.9.6

- A new machine-readable format exists for easy parsing by GPG, et al.
- Lots of bug squashing, documentation updates, etc.
- Temporary fixes for corruption that occurs with subkeys. (The packets
  are dropped instead of being added in a corrupt format.)
  READ THE FILE "UPGRADING" FOR INFORMATION ON UPGRADING TO 0.9.6!!!

***** Changes in version 0.9.5

- Allow 16-digit and 40-digit keyids/fingerprints.  Currently only the
  low 8 digits are used, but this should smooth a transition to a
  future pksd that allows long keyid or fingerprint searches.
- Nickolai Zeldovich's fix for garbage mixed in with large keys.
- Add missing transaction hooks (patch2)
- Make checkpoint code retry if necessary (patch2)
- Change the way memory is allocated in some places to reduce the
number of extra calls to malloc and free (patch2)
- Fix a bug in db 2.7.5 which causes db_recover to fail under some
circumstances. (patch2)
- Stopped flooding caused by multiple revokation certs (flood.patch)
- In kd_search, fixed bug in using maxid when keyid not found
(kdsearch_error.patch)
- WWW key submissions disabled by default (JHPatch1)
- nicer formatting of SHA-1 fingerprints (JHPatch1)
- pksdctl() usage page improvements (JHPatch1)
- manual page improvements (JHPatch1)
- Fixed buffer overflow from long arguments to search in HDP requests,
as mentioned on bugtraq (kd_search by Jason Harris)
- Fixed problem of randomly not sending updated sigs to syncsites due
to improperly initialized memory. (patch_pf20020615)
- Added compile-time support for X509 recognition (x509.patch)
- Fixed makefiles to more support rpm build roots (pks-bldroot.patch)
- Changed install to put db2 in own subdir (pks-db2-install.patch)
- Added libwrap support (pks-libwrap.patch)
- Added sample init script (pks-rc.patch)
- Beautified some of the web output (pks-beauty.patch)
- Fixed a buffer overflow (from Teun Nijssen's source tree)
- Client connection aborts are now handled properly on newer versions
of Solaris (from Teun Nijssen's source tree - original change by
Xander Jansen of SURFnet)
- Added checkpointing after so many transactions. (patch3)
- Added more logging to kvcv.c (patch3)
- Added some messages regarding the non-support of X.509. (patch3)
- Added max_last_reply_keys configuration option (patch3)
- The spec file was radically updated for the upcoming pks-0.9.5rc1.
  o The database is now stored in /var/lib/pks instead of /home/keyserver
  o The PREFIX is now /usr instead of /usr/local
  o For more details, see the %changelog that is within the spec file.

***** Changes in version 0.9.4

- Converted to Sleepycat Software Berkeley DB version 2.7.5.
- mmap()d input files instead of reading them when possible.
- The database may be split into multiple smaller files.
- When a key does not parse, return a soft error to the user and
continue processing the input keyblock.
- When extending lists in the database, do so in chunks, rather than
one entry at a time.  This improves performance when adding keys.
- Allow the administrator to limit the maximum duration of the LAST
mail server command.
- Allow the administrator to limit the number of keys returned by
index, verbose index, get, and last requests.
- Use the log, lock, and transaction features of Berkeley DB.

***** Changes in version 0.9.2

- Added support for pgp 5's DSA and ElGamal keys.
- Added support to short-circuit the normal output paths in
pksclient, and instead output indexes and binary keys directly to
stdout.  This saves a lot of virtual memory.
- Removed many gmake-isms from the Makefile.
- In the index pages from the web server, links in the User ID look
up by KeyID instead of name, to eliminate ambiguity.
- Fixed defunct process leak on some platforms, notably solaris.
- Check for and, if necessary, use bsd4.2/ultrix 2-argument openlog.
- The help files now must have a (possibly empty) MIME header block.
This allows help files to use non-ascii character sets.
- keyrings returned by the mail server have content-type
application/pgp-keys, as per rfc2015.
- fixed a double-memory-free.
- pksd is more careful about shutting down cleanly when it receives a
non-fatal signal (INT, QUIT, HUP, TERM).


pksclient.c
const char rcsid_pksclient_c[] = "$Id: pksclient.c,v 1.1.1.1 2002/09/04 20:48:53 dtype Exp $";

/* 
 * Copyright (c) 1996, 1997, 1998, 1999, Marc Horowitz.  All rights reserved.
 * See the LICENSE file in the release for redistribution information.
 */


#include <sys/types.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>

#include "database.h"
#include "globals.h"
#include "util.h"

void usage(char *argv0)
{
   fprintf(stderr,
	   "%s /db/path create [num_files]\n"
	   "%s /db/path recover\n"
	   "%s /db/path add filename [flags]\n"
	   "%s /db/path get userid [flags]\n"
	   "%s /db/path index userid [flags]\n"
	   "%s /db/path since time [flags]\n"
	   "%s /db/path delete userid [flags]\n"
	   "%s /db/path disable userid [flags]\n",
	   argv0, argv0, argv0, argv0, argv0, argv0, argv0, argv0);
   exit(1);
}

int main(int argc, char *argv[])
{
   char *dir, *op;
   unsigned char *ret;
   long retlen;

   if (argc < 3)
      usage(argv[0]);

   dir = argv[1];
   op = argv[2];

   ret = NULL;

   log_terminal = 1;

   if (strcmp(op, "create") == 0) {
      int num_files;

      num_files = argv[3]?atoi(argv[3]):3;

      if (!kd_create(dir, num_files, (char **) &ret)) {
	 fprintf(stderr, "database create failed: %s\n", ret);
	 free(ret);
	 exit(1);
      }

      printf("Database files created in %s\n", dir);
   } else if (strcmp(op, "recover") == 0) {
      if (!kd_open(dir, KD_OPEN_RECOVER, (char **) &ret)) {
	 fprintf(stderr, "database recovery failed: %s\n", ret);
	 free(ret);
	 exit(1);
      }

      printf("Database files recovered in %s\n", dir);
   } else if (strcmp(op, "add") == 0) {
      unsigned char *ret, *newkeys;
      long retlen, newkeyslen;
      char *fn, *opts;
      int donew = 0, flags = 0, oflags = 0;
      FILE *f;
      xfilecontents infile;

      if ((fn = argv[3]) == NULL)
	 usage(argv[0]);

      if ((opts = argv[4])) {
	 if (strchr(opts, 'n'))
	    donew = 1;
	 if (strchr(opts, 'd'))
	    flags |= KD_ADD_NO_STRIP_DISABLED;
	 if (strchr(opts, 't'))
	    oflags |= KD_OPEN_NOTXN;
      }

      if ((f = fopen(fn, "r")) == NULL) {
	 perror("opening file");
	 exit(1);
      }

      if (!xfilecontents_get(&infile, f)) {
	 fprintf(stderr, "%s: error getting input file %s contents\n",
		 argv[0], fn);
	 exit(1);
      }

      fclose(f);

      if (!kd_open(dir, oflags, (char **) &ret)) {
	 fprintf(stderr, "database open failed: %s\n", ret);
	 free(ret);
	 exit(1);
      }

      if (!kd_add(infile.buf, infile.len, KD_ADD_VERBOSE | flags,
		  &ret, &retlen,
		  donew?&newkeys:NULL, donew?&newkeyslen:NULL)) {
	 fprintf(stderr, "database add failed: %.*s\n", (int) retlen, ret);
	 free(ret);
	 xfilecontents_free(&infile);
	 kd_close();
	 exit(1);
      }

      xfilecontents_free(&infile);

      fwrite((void *) ret, (size_t) retlen, 1, stderr);
      if (donew)
	 fwrite((void *) newkeys, (size_t) newkeyslen, 1, stdout);
   } else if (strcmp(op, "get") == 0) {
      char *userid, *opts;
      int flags = 0, oflags = KD_OPEN_READONLY;

      if ((userid = argv[3]) == NULL)
	 usage(argv[0]);

      if ((opts = argv[4])) {
	 if (strchr(opts, 'e'))
	    flags |= KD_GET_EXACT;
	 if (strchr(opts, 'a'))
	    flags |= KD_GET_ALL;
	 if (strchr(opts, 'b'))
	    flags |= KD_GET_BINARY;
	 if (strchr(opts, 'i'))
	    flags |= KD_GET_IGNORE_ERRORS;
	 if (strchr(opts, 'd'))
	    flags |= KD_GET_RETURN_DISABLED;
	 if (strchr(opts, 's'))
	    flags |= KD_GET_STDOUT;
	 if (strchr(opts, 't'))
	    oflags |= KD_OPEN_NOTXN;
      }

      if (!kd_open(dir, oflags, (char **) &ret)) {
	 fprintf(stderr, "database open failed: %s\n", ret);
	 free(ret);
	 exit(1);
      }

      if (!kd_get((unsigned char *) userid, (long) strlen(userid),
		  flags, -1, &ret, &retlen)) {
	 fprintf(stderr, "database get failed: %.*s\n", (int) retlen, ret);
	 free(ret);
	 kd_close();
	 exit(1);
      }

      fwrite((void *) ret, (size_t) retlen, 1, stdout);
   } else if (strcmp(op, "index") == 0) {
      char *userid, *opts;
      int flags = 0, oflags = KD_OPEN_READONLY;

      if ((userid = argv[3]) == NULL)
	 usage(argv[0]);

      if ((opts = argv[4])) {
	 if (strchr(opts, 'v'))
	    flags |= KD_INDEX_VERBOSE;
	 if (strchr(opts, 'f'))
	    flags |= KD_INDEX_FINGERPRINT;
	 if (strchr(opts, 'e'))
	    flags |= KD_INDEX_EXACT;
	 if (strchr(opts, 'a'))
	    flags |= KD_INDEX_ALL;
	 if (strchr(opts, 'i'))
	    flags |= KD_INDEX_IGNORE_ERRORS;
	 if (strchr(opts, 's'))
	    flags |= KD_INDEX_STDOUT;
	 if (strchr(opts, 't'))
	    oflags |= KD_OPEN_NOTXN;
      }

      if (!kd_open(dir, oflags, (char **) &ret)) {
	 fprintf(stderr, "database open failed: %s\n", ret);
	 free(ret);
	 exit(1);
      }

      if (!kd_index((unsigned char *) userid, (long) strlen(userid),
		    flags, -1, &ret, &retlen)) {
	 fprintf(stderr, "database index failed: %.*s\n", (int) retlen, ret);
	 free(ret);
	 kd_close();
	 exit(1);
      }

      fwrite((void *) ret, (size_t) retlen, 1, stdout);
   } else if (strcmp(op, "since") == 0) {
      time_t since, last;
      char *opts;
      int flags = 0, oflags = KD_OPEN_READONLY;

      if (argv[3] == NULL)
	 usage(argv[0]);

      since = atoi(argv[3]);

      if ((opts = argv[4])) {
	 if (strchr(opts, 'b'))
	    flags |= KD_SINCE_BINARY;
	 if (strchr(opts, 'r')) {
	    time_t now;
	    now = time(NULL);
	    since = now - since;
         }
	 if (strchr(opts, 't'))
	    oflags |= KD_OPEN_NOTXN;
      }

      if (!kd_open(dir, oflags, (char **) &ret)) {
	 fprintf(stderr, "database open failed: %s\n", ret);
	 free(ret);
	 exit(1);
      }

      if (!kd_since(since, flags, -1, &last, &ret, &retlen)) {
	 fprintf(stderr, "database since failed: %.*s\n", (int) retlen, ret);
	 free(ret);
	 kd_close();
	 exit(1);
      }

      fprintf(stderr, "Last key added at %ld\n\n", (long) last);

      fwrite((void *) ret, (size_t) retlen, 1, stdout);
   } else if (strcmp(op, "delete") == 0) {
      char *userid, *opts;
      int oflags = 0;

      if ((userid = argv[3]) == NULL)
	 usage(argv[0]);

      if ((opts = argv[4])) {
	 if (strchr(opts, 't'))
	    oflags |= KD_OPEN_NOTXN;
      }

      if (!kd_open(dir, oflags, (char **) &ret)) {
	 fprintf(stderr, "database open failed: %s\n", ret);
	 free(ret);
	 exit(1);
      }

      if (!kd_delete((unsigned char *) userid, (long) strlen(userid),
		     0, &ret, &retlen)) {
	 fprintf(stderr, "database delete failed: %.*s\n", (int) retlen, ret);
	 free(ret);
	 kd_close();
	 exit(1);
      }

      fwrite((void *) ret, (size_t) retlen, 1, stdout);
   } else if (strcmp(op, "disable") == 0) {
      char *userid, *opts;
      int flags = 0, oflags = 0;

      if ((userid = argv[3]) == NULL)
	 usage(argv[0]);

      if ((opts = argv[4])) {
	 if (strchr(opts, 'c'))
	    flags |= KD_DISABLE_CLEAR;
	 if (strchr(opts, 't'))
	    oflags |= KD_OPEN_NOTXN;
      }

      if (!kd_open(dir, oflags, (char **) &ret)) {
	 fprintf(stderr, "database open failed: %s\n", ret);
	 free(ret);
	 exit(1);
      }

      if (!kd_disable((unsigned char *) userid, (long) strlen(userid),
		      flags, &ret, &retlen)) {
	 fprintf(stderr, "database disable failed: %.*s\n", (int) retlen, ret);
	 free(ret);
	 kd_close();
	 exit(1);
      }

      fwrite((void *) ret, (size_t) retlen, 1, stdout);
   } else {
      usage(argv[0]);
   }

   if (ret)
      free(ret);

   kd_close();

   exit(0);
}
